sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -99,180 +108,337 @@ object Hash {
+              case Value.ValueText(value) =>
+                if (value.nonEmpty)
+                  discard(addField.addString(value))
+              case Value.ValueTimestamp(value) =>
+                if (value.micros != 0)
+                  discard(addField.addLong(value.micros))
+              case Value.ValueDate(value) =>
+                if (value.days != 0)
+                  discard(addField.addInt(value.days))
+              case Value.ValueParty(value) =>
+                // No default value for party
+                discard(addField.addString(value))
+              case Value.ValueBool(value) =>
+                if (value)
+                  discard(addField.addByte(1.toByte))
+              case Value.ValueUnit =>
+              // We never write unit
+            }
+          case Value.ValueRecord(_, fields) =>
+            // No default value for records
+            discard(addField.addRecord(fields))
+          case Value.ValueVariant(_, variant, value) =>
+            // No default value for variant
+            discard(addField.addVariant(variant, value))
+          case Value.ValueContractId(value) =>
+            // No default value for contractId
+            discard(addField.addCid(value))
+          case Value.ValueList(values) =>
+            if (values.nonEmpty)
+              discard(addField.addList(values))
+          case Value.ValueOptional(value) =>
+            value match {
+              case Some(value) =>
+                discard(addField.addTypedValue(value))
+              case None =>
+            }
+          case Value.ValueTextMap(value) =>
+            if (!value.isEmpty)
+              discard(addTextMap(value))
+          case Value.ValueGenMap(entries) =>
+            if (entries.nonEmpty)
+              discard(addGenMap(entries))
+        }
+      }
+      // This delimits the end of the record.
+      // Note it does not collide with the first byte of field numbers as those are always positives.
+      addByte(0xff.toByte)
+    }
+
+  }
+
   // The purpose of a hash serves to avoid hash collisions due to equal encodings for different objects.
   // Each purpose should be used at most once.
-  private[crypto] case class Purpose(id: Byte)
+  private[crypto] class Purpose private (val id: Byte)
 
   private[crypto] object Purpose {
-    val Testing = Purpose(1)
-    val ContractKey = Purpose(2)
-    val MaintainerContractKeyUUID = Purpose(4)
-    val PrivateKey = Purpose(3)
-    val ContractInstance = Purpose(5)
-    val ChangeId = Purpose(6)
+    val Testing = new Purpose(1)
+    val ContractKey = new Purpose(2)
+    val MaintainerContractKeyUUID = new Purpose(4)
+    val PrivateKey = new Purpose(3)
+    val ContractInstance = new Purpose(5)
+    val ChangeId = new Purpose(6)
+  }
+
+  private class Version private (val id: Byte)
+
+  private object Version {
+    val Legacy = new Version(0) // LF 1.x to LF 2.1
+    val UpgradeFriendly = new Version(1) // from LF 2.1
   }
```

Comment:
Add a new version for the new scheme to prevent any collision between them. 
I forgot that in my initial commit.